File: beecrowd/2910 - Desativação de Rodovias.cpp
#include <bits/stdc++.h>

using namespace std;

using Edge = tuple<size_t, int, int>;
using Adj = vector<vector<Edge>>;

vector<long long> dijkstra(const Adj& adj) {
    using pq_type = pair<long long, size_t>;

    const auto n = adj.size();
    constexpr auto INF = numeric_limits<long long>::max();
    
    vector<long long> dist(n, INF);
    priority_queue<pq_type, vector<pq_type>, greater<pq_type>> pq;

    pq.emplace(0LL, 0);
    dist[0] = 0;

    while(!pq.empty()) {
        auto [cost, u] = pq.top();

        pq.pop();

        if(dist[u] != cost)
            continue;

        for(auto [v, l, c] : adj[u]) {
            auto new_cost = cost + l;

            if(dist[v] <= new_cost)
                continue;

            dist[v] = new_cost;
            pq.emplace(new_cost, v);
        }
    }

    return dist;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    size_t n, m;

    cin >> n >> m;

    Adj adj(n);

    for(size_t i = 0; i < m; ++i) {
        size_t a, b;
        int l, c;

        cin >> a >> b >> l >> c;

        --a, --b;

        adj[a].emplace_back(b, l, c);
        adj[b].emplace_back(a, l, c);
    }

    auto cost = dijkstra(adj);

    vector<int> min_in_edge(n, INT_MAX);

    min_in_edge[0] = 0;

    for(size_t u = 0; u < n; ++u) {
        for(auto [v, l, c] : adj[u]) {
            if(cost[v] == cost[u] + l)
                min_in_edge[v] = min(min_in_edge[v], c);
        }
    }

    auto total_repair_cost = accumulate(min_in_edge.begin(), min_in_edge.end(), 0LL);

    cout << total_repair_cost << '\n';

    return 0;
}

File: beecrowd/2910 - Desativação de Rodovias.rs
#![allow(dead_code)]
use std::{
    cmp::Reverse,
    collections::BinaryHeap,
    io::{self, Read, Write},
    writeln,
};

struct Scanner {
    buffer: Vec<u8>,
    index: usize,
}

impl Scanner {
    fn new() -> Self {
        let mut input = Vec::new();

        io::stdin().read_to_end(&mut input).unwrap();

        Self {
            buffer: input,
            index: 0,
        }
    }

    fn has_next(&mut self) -> bool {
        while self.index < self.buffer.len() && self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        self.index < self.buffer.len()
    }

    fn next<T: std::str::FromStr>(&mut self) -> T {
        if !self.has_next() {
            panic!("End Of File");
        }

        let start = self.index;

        while self.index < self.buffer.len() && !self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        std::str::from_utf8(&self.buffer[start..self.index])
            .unwrap()
            .parse::<T>()
            .ok()
            .unwrap()
    }
}

type Edge = (usize, i32, i32);
type Adj = Vec<Vec<Edge>>;

fn dijkstra(adj: &Adj) -> Vec<i64> {
    let n = adj.len();

    let mut cost = vec![i64::MAX; n];
    let mut pq = BinaryHeap::new();

    cost[0] = 0;
    pq.push(Reverse((0, 0)));

    while let Some(Reverse((cst, u))) = pq.pop() {
        if cst != cost[u] {
            continue;
        }

        for &(v, length, _) in &adj[u] {
            let total_cost = cst + length as i64;

            if cost[v] > total_cost {
                cost[v] = total_cost;
                pq.push(Reverse((total_cost, v)));
            }
        }
    }

    cost
}

fn main() {
    let mut writer = io::BufWriter::with_capacity(1 << 20, io::stdout());

    let mut sc = Scanner::new();

    let n = sc.next::<usize>();
    let m = sc.next::<usize>();

    let mut adj = vec![vec![]; n];

    for _ in 0..m {
        let u = sc.next::<usize>() - 1;
        let v = sc.next::<usize>() - 1;
        let length = sc.next::<i32>();
        let cost = sc.next::<i32>();

        adj[u].push((v, length, cost));
        adj[v].push((u, length, cost));
    }

    let dist = dijkstra(&adj);
    let mut min_in_edge = vec![i32::MAX; n];

    min_in_edge[0] = 0;

    for (u, neighbours) in adj.iter().enumerate() {
        for &(v, length, cost) in neighbours {
            if dist[v] == dist[u] + length as i64 {
                min_in_edge[v] = std::cmp::min(min_in_edge[v], cost);
            }
        }
    }

    let total_repair_cost = min_in_edge.iter().map(|c| *c as i64).sum::<i64>();

    writeln!(writer, "{}", total_repair_cost).ok();
}
